### PR TITLE
Thai/project managers

### PR DESCRIPTION
**Pull request by athai**

### CODE REVIEW

File: tests/projects.js [hunk 1]
```diff
@@ -281,6 +281,24 @@ module.exports = function(expect, request, baseUrl) {
...
+      requestOptions.form.auth.username = 'patcht';
+      requestOptions.form.auth.password = 'drowssap';
+      requestOptions.form.object = copyJsonObject(patchedProject);
+
+      request.post(requestOptions, function(err, res, body) {
+        expect(res.statusCode).to.equal(401);
+
+        expect(body.error).to.equal('Authorization failure');
+        expect(body.text).to.equal('patcht is not authorized to patch' + 
+          'this project');
```
Comment (tschuy) [inline tests/projects.js]: Change this to ``to patch project [name]``

**Comment by tschuy** [comment]
+1

File: src/projects.js [hunk 2]
```diff
@@ -326,130 +326,137 @@ module.exports = function(app) {
...
-              return slug.name;
-            });
-
-            const err = errors.errorSlugsAlreadyExist(overlappingSlugs);
+      // grab the project's name
+      knex('projects').first('name').where('id', '=', projectIdQuery)
+      .then(function(projectName) {
+        // access userroles table and check if the user is participating in
+        // the project
+        knex('userroles').where({user: user.id, project: projectIdQuery})
```
Comment (LyonesGamer) [inline src/projects.js]: You could make this a little more efficient by selecting id and name in the first query and then using that id in this query.

File: src/projects.js [hunk 3]
```diff
@@ -326,130 +326,137 @@ module.exports = function(app) {
...
+        .then(function(roles) {
+          if (roles.length === 0 || roles[0].manager === false) {
+            const err = errors.errorAuthorizationFailure(user.username,
+              'patch ' + projectName.name);
```
Comment (LyonesGamer) [inline src/projects.js]: I'd word this as "make changes to ..."

File: src/projects.js [hunk 4]
```diff
@@ -326,130 +326,137 @@ module.exports = function(app) {
...
           }
+          // retrieves the project from the database, selecting the project
+          // where its ID matches the slug's project (the projectIdQuery).
+
+          // also makes the owner field the username so it can be checked, and
+          // puts the ownerId into the ownerId field.
+          knex('projects').first().select('projects.id as id',
+          'projects.name as name', 'projects.uri as uri',
+          'users.username as owner', 'users.id as ownerId')
+          .where('projects.id', '=', projectIdQuery)
```
Comment (LyonesGamer) [inline src/projects.js]: Actually, you could just make this the query above. It'd be a bit extra overhead for cases where the person can't modify the project but one fewer query for when they can.

Comment (athai) [inline src/projects.js]: Which query are you referring to when you say "the query above"?

Comment (LyonesGamer) [inline src/projects.js]: The one I commented on. So right now you're doing "get project name -> get and check user role -> get whole project". You could just do "get whole project -> get and check user role".

File: src/projects.js [hunk 5]
```diff
@@ -332,124 +332,127 @@ module.exports = function(app) {
...
-          });
 
-          if (overlappingSlugs.length) {
-            overlappingSlugs = overlappingSlugs.map(function(slug) {
-              return slug.name;
-            });
-
-            const err = errors.errorSlugsAlreadyExist(overlappingSlugs);
+        // access userroles, check if user is participating in project
+        knex('userroles').where({user: user.id, project: projectIdQuery})
```
Comment (LyonesGamer) [inline src/projects.js]: Since you now have the id as ``project[0].id``, you can use that here instead of the subquery.

**Comment by LyonesGamer** [comment]
+1

**Comment by Kennric** [comment]
Looks good to me. +1

**Pull request merged**